include/bout/region.hxx

```diff
@@ -0,0 +1,236 @@
+ *
+ **************************************************************************/
+
+#ifndef __REGION_H__
+#define __REGION_H__
+
+#include <type_traits>
+#include <vector>
+#include <utility>
+
+#include "bout/openmpwrap.hxx"
+
+/*
+ * The MAXREGIONBLOCKSIZE value can be tuned to try to optimise
+ * performance on specific hardware. It determines what the largest
+ * contiguous block size can be. As we hope the compiler will vectorise
+ * the access to these contiguous blocks, the optimal MAXREGIONBLOCKSIZE 
+ * is likely related to the vector size etc.
+ */
+#ifndef MAXREGIONBLOCKSIZE
+#define MAXREGIONBLOCKSIZE 64
+#endif
+#if MAXREGIONBLOCKSIZE <= 0
+#error "In region.hxx must set MAXREGIONBLOCKSIZE > 0"
+#endif
+
+//The following loops vectorise well well index is type int, needs testing
+//with current approach where index is ind2D or ind3D
+#define BLOCK_REGION_LOOP_SERIAL(region, index, ...)			\
+  {const auto blocks = region.getBlocks();				\
+  for(auto block = blocks.begin(); block < blocks.end() ; ++block){ \
+  for(auto index = (*block).first ; index <= (*block).second; ++index){ \
+  __VA_ARGS__ \
+    }}}
+
+#define BLOCK_REGION_LOOP(region, index, ...)			\
+  {const auto blocks = region.getBlocks();			\
+  BOUT_OMP(parallel for)				    \
+  for(auto block = blocks.begin(); block < blocks.end() ; ++block){ \
+  for(auto index = (*block).first ; index <= (*block).second; ++index){ \
+  __VA_ARGS__ \
+    }}}
+
+class specificInd{
+public:
+  int ind;
+  specificInd(){};
+  specificInd(int i): ind(i){};
+  specificInd& operator++(){++ind;return *this;}
+  specificInd& operator++(int){++ind;return *this;}
+  bool operator==(const specificInd &x) const {
+    return ind == x.ind;
+  }
+  bool operator!=(const specificInd &x) const {
+    return ind != x.ind;
+  }
+  bool operator>(const specificInd &x) const {
+    return ind > x.ind;
+  }
+  bool operator>=(const specificInd &x) const {
+    return ind >= x.ind;
+  }
+  bool operator<(const specificInd &x) const {
+    return ind < x.ind;
+  }
+  bool operator<=(const specificInd &x) const {
+    return ind <= x.ind;
+  }
+};
+//Make identical subclasses with different names
+class ind3D: public specificInd{
+public:
+  ind3D(): specificInd(){};
+  ind3D(int i): specificInd(i){};
+  //Note operator= from base class is always hidden
+  //by implicit method so have to be explicit
+  ind3D& operator=(const int& i){ind=i;return *this;}
+};
+class ind2D: public specificInd{
+public:
+  ind2D(): specificInd(){};
+  ind2D(int i): specificInd(i){};
+  ind2D& operator=(const int& i){ind=i;return *this;}
+};
+
+template<typename T = ind3D>
+class Region {
+  //Following prevents a Region being created with anything other
+  //than ind2D or ind3D as template type
+  static_assert(std::is_base_of<ind2D,T>::value || std::is_base_of<ind3D,T>::value, "Region must be templated with either ind2D or ind3D");
+  
+public:
+  typedef T data_type;
```

Comment:
This isn't used anywhere, is it needed?

Reply:
It's not currently used but I thought it might be useful to be able to query what a given Region is templated with (say if we start using `auto` in places like return values with C++14).